File: go/PathSum/PathSum.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(logN)
func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	newTarget := targetSum - root.Val
	if newTarget == 0 && root.Left == nil && root.Right == nil {
		return true
	}

	left := hasPathSum(root.Left, newTarget)
	right := hasPathSum(root.Right, newTarget)

	return left || right
}

func main() {
	left := &TreeNode{Val: 2}
	right := &TreeNode{Val: 3}
	root := &TreeNode{Val: 1, Left: left, Right: right}

	fmt.Println(hasPathSum(root, 3))
}

File: python/pathSum.py
from typing import Union


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


# Time -> 0(N) Space -> 0(logN)
def hasPathSum(root: Union[TreeNode, None], targetSum: int) -> bool:
    if root is None:
        return False

    newTarget = targetSum - root.val
    if newTarget == 0 and root.left is None and root.right is None:
        return True

    left = hasPathSum(root.left, newTarget)
    right = hasPathSum(root.right, newTarget)

    return left or right


left = TreeNode(val=2)
right = TreeNode(val=3)
root = TreeNode(1, left, right)

print(hasPathSum(root, 5))
